include/swift/RemoteInspection/ReflectionContext.h

```diff
@@ -799,6 +807,297 @@ class ReflectionContext
     }
   }
 
+  /// Parses metadata information from a WebAssembly image.
+  ///
+  ///
+  /// \param[in] ImageStart
+  ///     A remote address pointing to the start of the image in the running
+  ///     process.
+  ///
+  /// \param[in] FileBuffer
+  ///     A buffer which contains the contents of the image's file
+  ///     in disk. If missing, all the information will be read using the
+  ///     instance's memory reader.
+  ///
+  /// \return The newly added reflection info ID if successful, \b std::nullopt
+  ///     otherwise.
+  std::optional<uint32_t>
+  readWasm(RemoteAddress ImageStart,
+           std::optional<llvm::sys::MemoryBlock> FileBuffer,
+           llvm::SmallVector<llvm::StringRef, 1> PotentialModuleNames) {
+    /// A WASM data segment. The reflection metadata "sections" are DATA
+    /// segments.
+    struct Segment {
+      RemoteAddress remoteAddr;
+      uint64_t offset;
+      uint64_t size;
+    };
+    std::map<std::string, Segment> sections;
+    std::vector<Segment> segments;
+    auto &reader = getReader();
+    RemoteAddress cursor = ImageStart;
+    cursor += sizeof(llvm::wasm::WasmMagic) + sizeof(llvm::wasm::WasmVersion);
+
+    /// Decode one byte and move the cursor.
+    auto decodeU8 = [&reader, &cursor](uint8_t &b) -> bool {
+      if (!reader.readInteger(cursor, 1, &b))
+        return false;
+      cursor += 1;
+      return true;
+    };
+
+    /// Decode 32-bit ULEB constants.
+    auto decodeULEB32 = [&](uint32_t &val) -> bool {
```

Comment:
Unfortunate that we have to reimplement these for the umpteenth time. Is there no code object format parsing code in LLVM to reuse here?